File: EdgeGodParlays/examples/expert_engine_odds.py
#!/usr/bin/env python3
"""
Enhanced Expert Engine Odds Client - Python Version
Combines your sample odds.py with EdgeGod rate limiting + expert engine integration points

Features:
- Built-in 429 error prevention (EdgeGod rate limiting)
- Expert filter integration hooks
- Time window filtering (commenceTimeFrom/commenceTimeTo)
- Best price detection
- Implied probability calculations
- Value threshold analysis
- Production-ready error handling
"""

import hashlib
import json
import os
import time
from dataclasses import dataclass
from datetime import datetime, timedelta

import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry


@dataclass
class ExpertFilter:
    """Configuration for expert engine filters"""

    min_implied_probability: float = 0.40  # 40% minimum
    max_implied_probability: float = 0.65  # 65% maximum
    min_value_threshold: float = 0.05  # 5% minimum edge
    preferred_markets: list[str] = None
    preferred_sports: list[str] = None
    time_window_hours: int = 24  # Next 24 hours

    def __post_init__(self):
        if self.preferred_markets is None:
            self.preferred_markets = ["h2h", "spreads", "totals"]
        if self.preferred_sports is None:
            self.preferred_sports = [
                "americanfootball_nfl",
                "basketball_nba",
                "soccer_epl",
            ]


class EdgeGodExpertOddsClient:
    """Enhanced Odds API client with expert engine integration"""

    def __init__(self, api_key: str, expert_filter: ExpertFilter = None):
        self.api_key = api_key
        self.expert_filter = expert_filter or ExpertFilter()
        self.base_url = "https://api.the-odds-api.com/v4"

        # EdgeGod rate limiting
        self.rate_limit = 25.0  # Conservative 25 req/sec
        self.min_interval = 1.0 / self.rate_limit
        self.last_request_time = 0.0
        self.cache = {}

        # Setup session with retry logic
        self.session = requests.Session()
        retry_strategy = Retry(
            total=3,
            status_forcelist=[429, 500, 502, 503, 504],
            backoff_factor=1,
            respect_retry_after_header=True,
        )
        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)

    def _wait_for_rate_limit(self):
        """EdgeGod rate limiting implementation"""
        now = time.time()
        time_since_last = now - self.last_request_time

        if time_since_last < self.min_interval:
            sleep_time = self.min_interval - time_since_last
            time.sleep(sleep_time)

        self.last_request_time = time.time()

    def _get_cache_key(self, url: str, params: dict) -> str:
        """Generate cache key for request"""
        cache_data = f"{url}_{json.dumps(params, sort_keys=True)}"
        return hashlib.md5(cache_data.encode()).hexdigest()

    def _make_request(self, url: str, params: dict) -> dict:
        """Make rate-limited API request with caching"""
        params = params.copy()
        params["apiKey"] = self.api_key

        # Check cache (15 min TTL)
        cache_key = self._get_cache_key(url, params)
        if cache_key in self.cache:
            cached_data, timestamp = self.cache[cache_key]
            if time.time() - timestamp < 900:  # 15 minutes
                print(f"✅ Cache hit for {url.split('/')[-1]}")
                return cached_data

        # Apply rate limiting
        self._wait_for_rate_limit()

        try:
            response = self.session.get(url, params=params, timeout=30)

            if response.status_code == 429:
                print("⚠️ Rate limit hit, applying backoff...")
                time.sleep(2)
                return self._make_request(url, params)

            response.raise_for_status()
            data = response.json()

            # Cache successful response
            self.cache[cache_key] = (data, time.time())

            return data

        except requests.exceptions.RequestException as e:
            print(f"❌ API request failed: {e}")
            raise

    def get_sports(self) -> list[dict]:
        """Get available sports - enhanced version of your original"""
        url = f"{self.base_url}/sports/"
        sports = self._make_request(url, {})

        # 🎯 EXPERT ENGINE INTEGRATION POINT
        if self.expert_filter.preferred_sports:
            sports = [s for s in sports if s["key"] in self.expert_filter.preferred_sports]
            print(f"🎯 Filtered to {len(sports)} preferred sports")

        return sports

    def get_odds(
        self,
        sport_key: str,
        regions="us",
        markets="h2h",
        odds_format="american",
        with_time_filter=True,
    ) -> list[dict]:
        """Get odds with expert engine enhancements"""
        url = f"{self.base_url}/sports/{sport_key}/odds"
        params = {
            "regions": regions,
            "markets": markets,
            "oddsFormat": odds_format,
        }

        # 🎯 EXPERT ENGINE TIME FILTERING
        if with_time_filter:
            now = datetime.utcnow()
            commence_from = now.isoformat() + "Z"
            commence_to = (
                now + timedelta(hours=self.expert_filter.time_window_hours)
            ).isoformat() + "Z"
            params["commenceTimeFrom"] = commence_from
            params["commenceTimeTo"] = commence_to
            print(f"🕒 Filtering events: next {self.expert_filter.time_window_hours} hours")

        odds_data = self._make_request(url, params)

        # 🎯 EXPERT ENGINE FILTERING
        filtered_odds = []
        for event in odds_data:
            expert_analysis = self.analyze_event_for_expert_engine(event)
            if expert_analysis["passes_filters"]:
                event["expert_analysis"] = expert_analysis
                filtered_odds.append(event)

        print(f"🎯 Expert filter: {len(filtered_odds)}/{len(odds_data)} events passed")
        return filtered_odds

    def analyze_event_for_expert_engine(self, event: dict) -> dict:
        """
        🎯 EXPERT ENGINE CORE ANALYSIS
        This is where you'd plug in your expert logic
        """
        analysis = {
            "passes_filters": False,
            "best_prices": {},
            "implied_probabilities": {},
            "value_opportunities": [],
            "recommended_bets": [],
        }

        try:
            # Find best prices across all bookmakers for each market
            for bookmaker in event.get("bookmakers", []):
                for market in bookmaker.get("markets", []):
                    market_key = market["key"]

                    # Skip if not in preferred markets
                    if market_key not in self.expert_filter.preferred_markets:
                        continue

                    for outcome in market.get("outcomes", []):
                        outcome_name = outcome["name"]
                        price = outcome["price"]

                        # Track best price for each outcome
                        key = f"{market_key}_{outcome_name}"
                        if (
                            key not in analysis["best_prices"]
                            or price > analysis["best_prices"][key]["price"]
                        ):
                            analysis["best_prices"][key] = {
                                "price": price,
                                "bookmaker": bookmaker["key"],
                                "market": market_key,
                                "outcome": outcome_name,
                            }

                        # Calculate implied probability
                        if price > 0:  # American odds
                            implied_prob = (
                                100 / (price + 100) if price > 0 else (-price) / (-price + 100)
                            )
                        else:
                            implied_prob = (-price) / (-price + 100)

                        analysis["implied_probabilities"][key] = implied_prob

                        # 🎯 EXPERT FILTER: Check if within probability range
                        if (
                            self.expert_filter.min_implied_probability
                            <= implied_prob
                            <= self.expert_filter.max_implied_probability
                        ):

                            # Calculate potential value (simplified - you'd use your model here)
                            estimated_true_prob = self._estimate_true_probability(event, outcome)
                            value = estimated_true_prob - implied_prob

                            if value >= self.expert_filter.min_value_threshold:
                                analysis["value_opportunities"].append(
                                    {
                                        "market": market_key,
                                        "outcome": outcome_name,
                                        "price": price,
                                        "implied_prob": implied_prob,
                                        "estimated_prob": estimated_true_prob,
                                        "value": value,
                                        "bookmaker": bookmaker["key"],
                                    }
                                )

                                analysis["recommended_bets"].append(
                                    {
                                        "confidence": ("HIGH" if value > 0.10 else "MEDIUM"),
                                        "bet_type": market_key,
                                        "selection": outcome_name,
                                        "odds": price,
                                        "value_edge": f"{value:.1%}",
                                        "bookmaker": bookmaker["key"],
                                    }
                                )

            # Event passes if it has value opportunities
            analysis["passes_filters"] = len(analysis["value_opportunities"]) > 0

        except Exception as e:
            print(f"⚠️ Error analyzing event {event.get('id', 'unknown')}: {e}")

        return analysis

    def _estimate_true_probability(self, event: dict, outcome: dict) -> float:
        """
        🎯 EXPERT ENGINE: Your probability model goes here
        This is a placeholder - replace with your actual model
        """
        # Placeholder logic - you'd implement your sophisticated model here
        # Consider factors like: team strength, historical performance, injuries, etc.

        # For now, return a simple estimate based on market consensus
        return 0.50  # Replace with your model

    def get_expert_recommendations(self, sports_list: list[str] = None) -> dict:
        """
        🎯 EXPERT ENGINE MAIN FUNCTION
        Get recommendations across multiple sports
        """
        sports_list = sports_list or self.expert_filter.preferred_sports
        all_recommendations = []

        print("🎯 Running Expert Engine Analysis...")
        print("=" * 50)

        for sport in sports_list:
            try:
                print(f"\n🏆 Analyzing {sport}...")
                odds_data = self.get_odds(
                    sport,
                    regions="us",
                    markets=",".join(self.expert_filter.preferred_markets),
                    odds_format="american",
                )

                for event in odds_data:
                    if "expert_analysis" in event and event["expert_analysis"]["recommended_bets"]:
                        event_info = {
                            "sport": sport,
                            "home_team": event["home_team"],
                            "away_team": event["away_team"],
                            "commence_time": event["commence_time"],
                            "recommendations": event["expert_analysis"]["recommended_bets"],
                        }
                        all_recommendations.append(event_info)

            except Exception as e:
                print(f"❌ Error processing {sport}: {e}")

        return {
            "total_events_analyzed": sum(
                len(self.get_odds(sport, with_time_filter=False)) for sport in sports_list
            ),
            "events_with_value": len(all_recommendations),
            "recommendations": all_recommendations,
            "filter_settings": {
                "min_probability": f"{self.expert_filter.min_implied_probability:.1%}",
                "max_probability": f"{self.expert_filter.max_implied_probability:.1%}",
                "min_value": f"{self.expert_filter.min_value_threshold:.1%}",
                "time_window": f"{self.expert_filter.time_window_hours} hours",
            },
        }


def main():
    """Enhanced main function with expert engine integration"""
    # Setup
    API_KEY = os.getenv("ODDS_API_KEY") or "YOUR_API_KEY"

    if API_KEY == "YOUR_API_KEY":
        print("❌ Please set ODDS_API_KEY environment variable")
        return

    # Configure expert filters
    expert_config = ExpertFilter(
        min_implied_probability=0.35,  # 35% minimum
        max_implied_probability=0.70,  # 70% maximum
        min_value_threshold=0.03,  # 3% minimum edge
        preferred_sports=["americanfootball_nfl", "basketball_nba"],
        time_window_hours=48,  # Next 48 hours
    )

    # Initialize enhanced client
    client = EdgeGodExpertOddsClient(API_KEY, expert_config)

    print("🎯 EdgeGod Expert Engine - Enhanced Odds Analysis")
    print("=" * 60)
    print("✅ Built-in 429 error prevention")
    print("✅ Expert filter integration")
    print("✅ Best price detection")
    print("✅ Value opportunity analysis")
    print("✅ Time window filtering")
    print("=" * 60)

    try:
        # Get expert recommendations
        recommendations = client.get_expert_recommendations()

        print("\n📊 EXPERT ENGINE RESULTS:")
        print(f"   📈 Events analyzed: {recommendations['total_events_analyzed']}")
        print(f"   🎯 Events with value: {recommendations['events_with_value']}")
        print(f"   ⚙️ Filter settings: {recommendations['filter_settings']}")

        if recommendations["recommendations"]:
            print("\n🏆 TOP RECOMMENDATIONS:")
            for i, rec in enumerate(recommendations["recommendations"][:5], 1):
                print(f"\n   {i}. {rec['home_team']} vs {rec['away_team']}")
                print(f"      🕒 {rec['commence_time']}")
                print(f"      🏆 {rec['sport']}")

                for bet in rec["recommendations"]:
                    print(
                        f"      💰 {bet['confidence']} confidence: {bet['selection']} @ {bet['odds']} "
                        f"(Edge: {bet['value_edge']}) via {bet['bookmaker']}"
                    )
        else:
            print("\n📋 No value opportunities found with current filters")
            print("💡 Try adjusting ExpertFilter settings for different results")

    except Exception as e:
        print(f"❌ Error: {e}")


if __name__ == "__main__":
    main()

File: EdgeGodParlays/examples/expert_engine_odds.js
#!/usr/bin/env node
/**
 * Enhanced Expert Engine Odds Client - Node.js Version
 * Combines your sample odds.js with EdgeGod rate limiting + expert engine integration
 * 
 * Features:
 * - Built-in 429 error prevention (EdgeGod rate limiting)
 * - Expert filter integration hooks
 * - Time window filtering (commenceTimeFrom/commenceTimeTo)
 * - Best price detection
 * - Implied probability calculations
 * - Value threshold analysis
 * - Production-ready async/await patterns
 */

const axios = require('axios');
const crypto = require('crypto');

/**
 * Expert filter configuration for the engine
 */
class ExpertFilter {
  constructor(config = {}) {
    this.minImpliedProbability = config.minImpliedProbability || 0.40;  // 40% minimum
    this.maxImpliedProbability = config.maxImpliedProbability || 0.65;  // 65% maximum
    this.minValueThreshold = config.minValueThreshold || 0.05;          // 5% minimum edge
    this.preferredMarkets = config.preferredMarkets || ['h2h', 'spreads', 'totals'];
    this.preferredSports = config.preferredSports || ['americanfootball_nfl', 'basketball_nba', 'soccer_epl'];
    this.timeWindowHours = config.timeWindowHours || 24;               // Next 24 hours
  }
}

/**
 * Enhanced Odds API client with expert engine integration
 */
class EdgeGodExpertOddsClient {
  constructor(apiKey, expertFilter = null) {
    this.apiKey = apiKey;
    this.expertFilter = expertFilter || new ExpertFilter();
    this.baseUrl = 'https://api.the-odds-api.com/v4';
    
    // EdgeGod rate limiting
    this.rateLimit = 25.0;  // Conservative 25 req/sec
    this.minInterval = 1000 / this.rateLimit; // milliseconds
    this.lastRequestTime = 0;
    this.cache = new Map();
    
    // Setup axios with retry logic
    this.setupAxiosRetry();
  }
  
  setupAxiosRetry() {
    // Configure axios interceptors for retry logic
    this.axiosInstance = axios.create({
      timeout: 30000,
    });
    
    // Response interceptor for 429 handling
    this.axiosInstance.interceptors.response.use(
      (response) => response,
      async (error) => {
        if (error.response?.status === 429) {
          console.log('⚠️ Rate limit hit, applying exponential backoff...');
          await this.sleep(2000); // 2 second backoff
          return this.axiosInstance.request(error.config);
        }
        return Promise.reject(error);
      }
    );
  }
  
  async sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
  
  async waitForRateLimit() {
    /**
     * EdgeGod rate limiting implementation
     */
    const now = Date.now();
    const timeSinceLast = now - this.lastRequestTime;
    
    if (timeSinceLast < this.minInterval) {
      const waitTime = this.minInterval - timeSinceLast;
      await this.sleep(waitTime);
    }
    
    this.lastRequestTime = Date.now();
  }
  
  getCacheKey(url, params) {
    const cacheData = url + JSON.stringify(params);
    return crypto.createHash('md5').update(cacheData).digest('hex');
  }
  
  async makeRequest(url, params = {}) {
    /**
     * Make rate-limited API request with caching
     */
    params = { ...params, apiKey: this.apiKey };
    
    // Check cache (15 min TTL)
    const cacheKey = this.getCacheKey(url, params);
    if (this.cache.has(cacheKey)) {
      const { data, timestamp } = this.cache.get(cacheKey);
      if (Date.now() - timestamp < 900000) { // 15 minutes
        console.log(`✅ Cache hit for ${url.split('/').pop()}`);
        return data;
      }
      this.cache.delete(cacheKey);
    }
    
    // Apply rate limiting
    await this.waitForRateLimit();
    
    try {
      const response = await this.axiosInstance.get(url, { params });
      const data = response.data;
      
      // Cache successful response
      this.cache.set(cacheKey, { data, timestamp: Date.now() });
      
      return data;
      
    } catch (error) {
      console.error(`❌ API request failed: ${error.message}`);
      throw error;
    }
  }
  
  async getSports() {
    /**
     * Get available sports - enhanced version of your original
     */
    const url = `${this.baseUrl}/sports/`;
    let sports = await this.makeRequest(url);
    
    // 🎯 EXPERT ENGINE INTEGRATION POINT
    if (this.expertFilter.preferredSports.length > 0) {
      sports = sports.filter(s => this.expertFilter.preferredSports.includes(s.key));
      console.log(`🎯 Filtered to ${sports.length} preferred sports`);
    }
    
    return sports;
  }
  
  async getOdds(sportKey, regions = 'us', markets = 'h2h', oddsFormat = 'american', withTimeFilter = true) {
    /**
     * Get odds with expert engine enhancements
     */
    const url = `${this.baseUrl}/sports/${sportKey}/odds`;
    const params = {
      regions,
      markets,
      oddsFormat,
    };
    
    // 🎯 EXPERT ENGINE TIME FILTERING
    if (withTimeFilter) {
      const now = new Date();
      const commenceFrom = now.toISOString();
      const commenceTo = new Date(now.getTime() + (this.expertFilter.timeWindowHours * 60 * 60 * 1000)).toISOString();
      params.commenceTimeFrom = commenceFrom;
      params.commenceTimeTo = commenceTo;
      console.log(`🕒 Filtering events: next ${this.expertFilter.timeWindowHours} hours`);
    }
    
    const oddsData = await this.makeRequest(url, params);
    
    // 🎯 EXPERT ENGINE FILTERING
    const filteredOdds = [];
    for (const event of oddsData) {
      const expertAnalysis = this.analyzeEventForExpertEngine(event);
      if (expertAnalysis.passesFilters) {
        event.expertAnalysis = expertAnalysis;
        filteredOdds.push(event);
      }
    }
    
    console.log(`🎯 Expert filter: ${filteredOdds.length}/${oddsData.length} events passed`);
    return filteredOdds;
  }
  
  analyzeEventForExpertEngine(event) {
    /**
     * 🎯 EXPERT ENGINE CORE ANALYSIS
     * This is where you'd plug in your expert logic
     */
    const analysis = {
      passesFilters: false,
      bestPrices: {},
      impliedProbabilities: {},
      valueOpportunities: [],
      recommendedBets: []
    };
    
    try {
      // Find best prices across all bookmakers for each market
      for (const bookmaker of (event.bookmakers || [])) {
        for (const market of (bookmaker.markets || [])) {
          const marketKey = market.key;
          
          // Skip if not in preferred markets
          if (!this.expertFilter.preferredMarkets.includes(marketKey)) {
            continue;
          }
          
          for (const outcome of (market.outcomes || [])) {
            const outcomeName = outcome.name;
            const price = outcome.price;
            
            // Track best price for each outcome
            const key = `${marketKey}_${outcomeName}`;
            if (!analysis.bestPrices[key] || price > analysis.bestPrices[key].price) {
              analysis.bestPrices[key] = {
                price: price,
                bookmaker: bookmaker.key,
                market: marketKey,
                outcome: outcomeName
              };
            }
            
            // Calculate implied probability
            let impliedProb;
            if (price > 0) { // American odds
              impliedProb = 100 / (price + 100);
            } else {
              impliedProb = (-price) / (-price + 100);
            }
            
            analysis.impliedProbabilities[key] = impliedProb;
            
            // 🎯 EXPERT FILTER: Check if within probability range
            if (impliedProb >= this.expertFilter.minImpliedProbability && 
                impliedProb <= this.expertFilter.maxImpliedProbability) {
              
              // Calculate potential value (simplified - you'd use your model here)
              const estimatedTrueProb = this.estimateTrueProbability(event, outcome);
              const value = estimatedTrueProb - impliedProb;
              
              if (value >= this.expertFilter.minValueThreshold) {
                analysis.valueOpportunities.push({
                  market: marketKey,
                  outcome: outcomeName,
                  price: price,
                  impliedProb: impliedProb,
                  estimatedProb: estimatedTrueProb,
                  value: value,
                  bookmaker: bookmaker.key
                });
                
                analysis.recommendedBets.push({
                  confidence: value > 0.10 ? 'HIGH' : 'MEDIUM',
                  betType: marketKey,
                  selection: outcomeName,
                  odds: price,
                  valueEdge: `${(value * 100).toFixed(1)}%`,
                  bookmaker: bookmaker.key
                });
              }
            }
          }
        }
      }
      
      // Event passes if it has value opportunities
      analysis.passesFilters = analysis.valueOpportunities.length > 0;
      
    } catch (error) {
      console.error(`⚠️ Error analyzing event ${event.id || 'unknown'}: ${error.message}`);
    }
    
    return analysis;
  }
  
  estimateTrueProbability(event, outcome) {
    /**
     * 🎯 EXPERT ENGINE: Your probability model goes here
     * This is a placeholder - replace with your actual model
     */
    // Placeholder logic - you'd implement your sophisticated model here
    // Consider factors like: team strength, historical performance, injuries, etc.
    
    // For now, return a simple estimate based on market consensus
    return 0.50; // Replace with your model
  }
  
  async getExpertRecommendations(sportsList = null) {
    /**
     * 🎯 EXPERT ENGINE MAIN FUNCTION
     * Get recommendations across multiple sports
     */
    sportsList = sportsList || this.expertFilter.preferredSports;
    const allRecommendations = [];
    
    console.log('🎯 Running Expert Engine Analysis...');
    console.log('='.repeat(50));
    
    for (const sport of sportsList) {
      try {
        console.log(`\n🏆 Analyzing ${sport}...`);
        const oddsData = await this.getOdds(
          sport,
          'us',
          this.expertFilter.preferredMarkets.join(','),
          'american'
        );
        
        for (const event of oddsData) {
          if (event.expertAnalysis && event.expertAnalysis.recommendedBets.length > 0) {
            const eventInfo = {
              sport: sport,
              homeTeam: event.home_team,
              awayTeam: event.away_team,
              commenceTime: event.commence_time,
              recommendations: event.expertAnalysis.recommendedBets
            };
            allRecommendations.push(eventInfo);
          }
        }
        
      } catch (error) {
        console.error(`❌ Error processing ${sport}: ${error.message}`);
      }
    }
    
    return {
      totalEventsAnalyzed: await this.getTotalEventsCount(sportsList),
      eventsWithValue: allRecommendations.length,
      recommendations: allRecommendations,
      filterSettings: {
        minProbability: `${(this.expertFilter.minImpliedProbability * 100).toFixed(1)}%`,
        maxProbability: `${(this.expertFilter.maxImpliedProbability * 100).toFixed(1)}%`,
        minValue: `${(this.expertFilter.minValueThreshold * 100).toFixed(1)}%`,
        timeWindow: `${this.expertFilter.timeWindowHours} hours`
      }
    };
  }
  
  async getTotalEventsCount(sportsList) {
    let total = 0;
    for (const sport of sportsList) {
      try {
        const odds = await this.getOdds(sport, 'us', 'h2h', 'american', false);
        total += odds.length;
      } catch (error) {
        // Continue on error
      }
    }
    return total;
  }
}

/**
 * Enhanced main function with expert engine integration
 */
async function main() {
  // Setup
  const API_KEY = process.env.ODDS_API_KEY || process.env.API_KEY || 'YOUR_API_KEY';
  
  if (API_KEY === 'YOUR_API_KEY') {
    console.error('❌ Please set ODDS_API_KEY environment variable');
    process.exit(1);
  }
  
  // Configure expert filters
  const expertConfig = new ExpertFilter({
    minImpliedProbability: 0.35,    // 35% minimum
    maxImpliedProbability: 0.70,    // 70% maximum
    minValueThreshold: 0.03,        // 3% minimum edge
    preferredSports: ['americanfootball_nfl', 'basketball_nba'],
    timeWindowHours: 48             // Next 48 hours
  });
  
  // Initialize enhanced client
  const client = new EdgeGodExpertOddsClient(API_KEY, expertConfig);
  
  console.log('🎯 EdgeGod Expert Engine - Enhanced Odds Analysis');
  console.log('='.repeat(60));
  console.log('✅ Built-in 429 error prevention');
  console.log('✅ Expert filter integration');
  console.log('✅ Best price detection');
  console.log('✅ Value opportunity analysis');
  console.log('✅ Time window filtering');
  console.log('='.repeat(60));
  
  try {
    // Get expert recommendations
    const recommendations = await client.getExpertRecommendations();
    
    console.log(`\n📊 EXPERT ENGINE RESULTS:`);
    console.log(`   📈 Events analyzed: ${recommendations.totalEventsAnalyzed}`);
    console.log(`   🎯 Events with value: ${recommendations.eventsWithValue}`);
    console.log(`   ⚙️ Filter settings: ${JSON.stringify(recommendations.filterSettings, null, 6)}`);
    
    if (recommendations.recommendations.length > 0) {
      console.log(`\n🏆 TOP RECOMMENDATIONS:`);
      recommendations.recommendations.slice(0, 5).forEach((rec, i) => {
        console.log(`\n   ${i + 1}. ${rec.homeTeam} vs ${rec.awayTeam}`);
        console.log(`      🕒 ${rec.commenceTime}`);
        console.log(`      🏆 ${rec.sport}`);
        
        rec.recommendations.forEach(bet => {
          console.log(`      💰 ${bet.confidence} confidence: ${bet.selection} @ ${bet.odds} ` +
                     `(Edge: ${bet.valueEdge}) via ${bet.bookmaker}`);
        });
      });
    } else {
      console.log('\n📋 No value opportunities found with current filters');
      console.log('💡 Try adjusting ExpertFilter settings for different results');
    }
    
  } catch (error) {
    console.error(`❌ Error: ${error.message}`);
  }
}

// Export for module use
module.exports = { EdgeGodExpertOddsClient, ExpertFilter };

// Run if called directly
if (require.main === module) {
  main().catch(console.error);
}